### PR TITLE
Initial layout for events page

### PR DESCRIPTION
**Pull request by alexzhang1618**
# Description

What changes did you make? List all distinct problems that this PR addresses. Explain any relevant
motivation or context.

- Laying out a prototype for an events page database
- Refactored Leaderboard.tsx to pull out lots of year stuff and make it common for both the Events and Leaderboard pages
- Kind of want to be able to adjust dropdown size for the common component (right now I've upscaled the search bar text to match, but I'd like to make both smaller)
- Would like feedback on the design/features that could be added

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) 

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [ ] locally on mobile - use https://ngrok.io to get a copy on a mobile device
- [ ] on the live deployment preview on Desktop.
- [ ] on the live deployment preview on Mobile.
- [ ] I have run and passed all new and existing Cypress tests. Add screenshots below.

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented my code's `src/lib` functions and commented hard to understand areas
      anywhere else.
- [x] My changes produce no new warnings.

# Screenshots

Screenshot 2023-12-30 at 11 27 48 PM
Screenshot 2023-12-30 at 11 28 03 PM
Screenshot 2023-12-30 at 11 28 22 PM

### CODE REVIEW

File: src/pages/events.tsx [hunk 1]
```diff
@@ -1,16 +1,170 @@
+import { Dropdown, PaginationControls, Typography } from '@/components/common';
+import EventDisplay from '@/components/events/EventDisplay';
+import { EventAPI } from '@/lib/api';
 import withAccessType from '@/lib/hoc/withAccessType';
-import { PermissionService } from '@/lib/services';
-import type { GetServerSideProps, NextPage } from 'next';
+import { CookieService, PermissionService } from '@/lib/services';
+import { PublicAttendance, PublicEvent } from '@/lib/types/apiResponses';
```
Comment (farisashai) [inline src/pages/events.tsx]:
```suggestion
import type { PublicAttendance, PublicEvent } from '@/lib/types/apiResponses';
```

File: src/pages/events.tsx [hunk 2]
```diff
@@ -1,16 +1,170 @@
+import { Dropdown, PaginationControls, Typography } from '@/components/common';
+import EventDisplay from '@/components/events/EventDisplay';
```
Comment (farisashai) [inline src/pages/events.tsx]:
```suggestion
import { EventDisplay } from '@/components/events';
```

File: src/pages/leaderboard.tsx [hunk 3]
```diff
@@ -83,14 +55,7 @@ const LeaderboardPage = ({ sort, leaderboard, user: { uuid } }: LeaderboardProps
...
 
-  const years = useMemo(() => {
-    const endYear = getEndYear();
-    const years = [];
-    for (let year = START_YEAR; year < endYear; year += 1) {
-      years.unshift({ value: String(year), label: `${year}–${year + 1}` });
-    }
-    return years;
-  }, []);
+  const years = useMemo(() => getYears(), []);
```
Comment (farisashai) [inline src/pages/leaderboard.tsx]:
```suggestion
  const years = useMemo(getYears, []);
```

File: src/pages/events.tsx [hunk 4]
```diff
@@ -1,16 +1,170 @@
...
+          <Dropdown
+            name="timeOptions"
+            ariaLabel="Filter events by time"
+            options={[
+              { value: 'upcoming', label: 'Upcoming' },
+              { value: 'past-week', label: 'Past week' },
+              { value: 'past-month', label: 'Past month' },
+              { value: 'past-year', label: 'Past year' },
+              { value: 'all-time', label: 'All time' },
+              '---',
```
Comment (farisashai) [inline src/pages/events.tsx:120]: what is this

**Review by farisashai — APPROVED** [review]
I think the search bar should be be white-space and case-insensitive and strip non alphanumeric characters so a search term like "aiintro toneuraL" would match the event "AI: Intro to Neural Networks"

File: src/pages/events.tsx (hunk 4 above)
Comment (SheepTester) [reply src/pages/events.tsx:120]: separator line

**Pull request merged**